File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        def helper(i,subset):
            if sum(subset)==target:
                res.append(subset[:])
                return
            if i>=len(candidates) or sum(subset)>target:
                return
            subset.append(candidates[i])
            helper(i,subset)
            subset.pop()
            helper(i+1,subset)
        #---------------------
        res=[]
        helper(0,[])
        return res

File: 39-combination-sum/combination-sum.java
class Solution {
    private void backtracking( List<List<Integer>> res, List<Integer> subset, int i,int sum,int[] candidates,int target){
        if(sum==target){
            res.add(new ArrayList<>(subset));
            return;
        }
        if(sum>target){
            return;
        }

        for(int j=i;j<candidates.length;j++){
            subset.add(candidates[j]);
            backtracking(res,subset,j,sum+candidates[j],candidates,target);
            subset.remove(subset.size()-1);
        }


    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res=new ArrayList<>();
        List<Integer> subset=new ArrayList<>();

        backtracking(res,subset,0,0,candidates,target);
        return res;


    }
}